File: 198.house-robber/solution3.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        N = len(nums)
        dp = [0] * (N + 2)
        for i in range(N):
            dp[i + 2] = nums[i] + dp[i]
            dp[i + 1] = max(dp[i + 1], dp[i])
        return max(dp[N], dp[N + 1])

File: 198.house-robber/solution3.kt
class Solution {
    fun rob(nums: IntArray): Int {
        
        val N = nums.count()
        if (N == 1) {
            return nums[0]
        }
        val dp = Array<Int>(N){0}
        dp[0] = nums[0]
        dp[1] = max(dp[0], nums[1])
        for (i in 2..N - 1) {
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])
        }
        
        // print(Arrays.toString(dp))
        return dp[N - 1]
    }
}